src/cls/rbd/cls_rbd.cc

```diff
@@ -4797,6 +4803,246 @@ int image_get_group(cls_method_context_t hctx,
   return 0;
 }
 
+/**
+ * Retrieve next group snapshot id.
+ *
+ * Input:
+ * none
+ *
+ * Output:
+ * @param uint64_t
+ * @return 0 on success, negative error code on failure
+ */
+int group_snap_next_seq(cls_method_context_t hctx,
+			bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_snap_next_seq");
+  bufferlist seqbl;
+  int r = cls_cxx_map_get_val(hctx, GROUP_SNAP_SEQ, &seqbl);
+  if (r < 0 && r != -ENOENT) {
+    return r;
+  }
+
+  bufferlist::iterator iter = seqbl.begin();
+  uint64_t snap_seq = 0;
+  try {
+    ::decode(snap_seq, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
+  ::encode(snap_seq, *out);
+
+  return 0;
+}
+
+/**
+ * Save initial snapshot record.
+ *
+ * Input:
+ * @param GroupSnapshot
+ *
+ * Output:
+ * @return 0 on success, negative error code on failure
+ */
+int group_snap_save(cls_method_context_t hctx,
+		    bufferlist *in, bufferlist *out)
+{
+  CLS_LOG(20, "group_snap_save");
+  cls::rbd::GroupSnapshot gs;
+  try {
+    bufferlist::iterator iter = in->begin();
+    ::decode(gs, iter);
+  } catch (const buffer::error &err) {
+    return -EINVAL;
+  }
+
```

Comment:
Need some basic validation of the provided group snapshot record. UUID empty, name empty, duplicate name, etc. Same comment applies to the update method.